python/nav/ipdevpoll/shadows/__init__.py

```diff
@@ -259,6 +266,38 @@ def _fix_binary_garbage(self):
                 setattr(self, attr, repr(value))
         self.clear_cached_objects()
 
+    def _detect_version_changes(self):
+        """
+        Detects if the software, hardware or firmware version changed for each device.
+
+        Saves this information in changed_versions in the Device instance.
+        """
+        old_device = self.get_existing_model()
+        self.changed_versions = ()
+        if old_device:
+            self.changed_versions = set(self.get_diff_attrs(old_device)).intersection(
+                (
+                    'hardware_version',
+                    'software_version',
+                    'firmware_version',
+                )
+            )
+
+    def cleanup(self, containers):
+        if self.changed_versions:
+            self._post_events_version_changes()
+
+    def _post_events_version_changes(self):
+        """Posts events for software, hardware or firmware changes."""
+        event = EventFactory('ipdevpoll', 'eventEngine', 'deviceNotice')
+        device = self.get_existing_model()
+        for version in self.changed_versions:
+            event.notify(
+                device=device,
+                subid=device.id,
+                alert_type=ALERT_TYPE_MAPPING[version],
+            ).save()
```

Comment:
I see some potential problems here, that might need some testing to iron out.

One bit that is interesting here is to add both the old version and the new version as variables to the event, so that they can both be referenced in the alert message ("device X changed software revision from X to Y").

Secondly, calling `self.get_existing_model()` *after* the save step will potentially just return the just saved object with the new values, meaning that the old version number might already have been lost.